src/fldeff_cut.c

```diff
@@ -1,19 +1,684 @@
+                    if (sHyperCutStruct[i].unk2[j] == 0)
+                        break;
+                    if (cutTiles[(u8)(sHyperCutStruct[i].unk2[j] - 1)] == FALSE)
+                    {
+                        tileCuttable = FALSE;
+                        break;
+                    }
+                    if (++j >= 2)
+                        break;
+                }
+
+                if (tileCuttable == TRUE)
+                {
+                    if (MapGridGetZCoordAt(x, y) == gPlayerFacingPosition.height)
+                    {
+                        u8 tileArrayId = ((sHyperCutStruct[i].y * 5) + 12) + (sHyperCutStruct[i].x);
+                        tileBehavior = MapGridGetMetatileBehaviorAt(x, y);
+                        if (MetatileBehavior_IsPokeGrass(tileBehavior) == TRUE
+                        || MetatileBehavior_IsAshGrass(tileBehavior) == TRUE)
+                        {
+                            gFieldCallback2 = FieldCallback_PrepareFadeInFromMenu;
+                            gPostMenuFieldCallback = FieldCallback_CutGrass;
+                            sHyperCutTiles[tileArrayId] = TRUE;
+                            ret = TRUE;
+                        }
+                        else
+                        {
+                            if (MetatileBehavior_IsCuttableGrass(tileBehavior) == TRUE)
+                                sHyperCutTiles[tileArrayId] = TRUE;
+                        }
+                    }
+                }
+            }
+
+            if (ret == TRUE)
+            {
+                gFieldCallback2 = FieldCallback_PrepareFadeInFromMenu;
+                gPostMenuFieldCallback = FieldCallback_CutGrass;
+            }
+        }
+
+        return ret;
+    }
+}
+
+static void FieldCallback_CutGrass(void)
+{
+    FieldEffectStart(FLDEFF_USE_CUT_ON_GRASS);
+    gFieldEffectArguments[0] = GetCursorSelectionMonId();
+}
+
+bool8 FldEff_UseCutOnGrass(void)
+{
+    u8 taskId = oei_task_add();
+
+    gTasks[taskId].data[8] = (u32)StartCutGrassFieldEffect >> 16;
+    gTasks[taskId].data[9] = (u32)StartCutGrassFieldEffect;
+    IncrementGameStat(GAME_STAT_USED_CUT);
+    return FALSE;
+}
+
+static void FieldCallback_CutTree(void)
+{
+    gFieldEffectArguments[0] = GetCursorSelectionMonId();
+    ScriptContext1_SetupScript(Route103_EventScript_290705);
+}
+
+bool8 FldEff_UseCutOnTree(void)
+{
+    u8 taskId = oei_task_add();
+
+    gTasks[taskId].data[8] = (u32)StartCutTreeFieldEffect >> 16;
+    gTasks[taskId].data[9] = (u32)StartCutTreeFieldEffect;
+    IncrementGameStat(GAME_STAT_USED_CUT);
+    return FALSE;
+}
+
+static void StartCutGrassFieldEffect(void)
+{
+    FieldEffectActiveListRemove(FLDEFF_USE_CUT_ON_GRASS);
+    FieldEffectStart(FLDEFF_CUT_GRASS);
+}
+
+bool8 FldEff_CutGrass(void)
+{
+    s16 x, y;
+    u8 tileBehavior;
+    u8 i;
+
+    for (i = 0, PlaySE(SE_W015), PlayerGetDestCoords(&gPlayerFacingPosition.x, &gPlayerFacingPosition.y); i < CUT_HYPER_AREA; i++)
```

Comment:
I think you should move the two functions out of the for statement (and if necessary, the i = 0 if it comes before those in the code)

Reply:
Copied directly from pokeruby.

Reply:
Still, if it can be done better than pokeruby...